File: design-pattern-python/design_pattern_python/abstract_factory/factory.py
from abc import ABCMeta, abstractmethod
from importlib import import_module
from typing import List


class Item(metaclass=ABCMeta):
    def __init__(self, caption) -> None:
        self.caption = caption

    @abstractmethod
    def makeHTML(self) -> str:
        pass


class Link(Item, metaclass=ABCMeta):
    def __init__(self, caption, url: str) -> None:
        self.url = url
        super().__init__(caption)


class Tray(Item, metaclass=ABCMeta):
    def __init__(self, caption: str) -> None:
        self.tray: List[Item] = []
        super().__init__(caption)

    def add(self, item: Item) -> None:
        self.tray.append(item)


class Page(metaclass=ABCMeta):
    def __init__(self, title: str, author: str) -> None:
        self.title = title
        self.author = author
        self.content: List[Item] = []

    def add(self, item: Item) -> None:
        self.content.append(item)

    def output(self, filename: str) -> None:
        with open(filename, "w") as f:
            f.write(self.makeHTML())
        print(filename + "を作成しました。")

    @abstractmethod
    def makeHTML(self) -> str:
        pass


class Factory(metaclass=ABCMeta):
    @classmethod
    def get_factory(cls, classname: str) -> "Factory":
        module, class_ = classname.rsplit(".", 1)
        return getattr(import_module(module), class_)()

    @abstractmethod
    def create_link(self, caption: str, url: str) -> Link:
        pass

    @abstractmethod
    def create_tray(self, caption: str) -> Tray:
        pass

    @abstractmethod
    def create_page(self, title: str, author: str) -> Page:
        pass

File: design-pattern-go/abstract_factory/factory/factory.go
package factory

import (
	"fmt"
	"log"
	"os"
)

/*****************************
 * Item
 ****************************/
type ItemInterface interface {
	MakeHTML() string
}

type Item struct {
	Caption string
}

/*****************************
 * Link
 ****************************/
type LinkInterface interface {
	ItemInterface
}

type Link struct {
	Item
	Url string
}

func NewLink(caption, url string) *Link {
	return &Link{
		Item: Item{
			caption,
		},
		Url: url,
	}
}

/*****************************
 * Tray
 ****************************/
type TrayInterface interface {
	ItemInterface
	Add(item ItemInterface)
}

type Tray struct {
	Item
	Tray []ItemInterface
}

func NewTray(caption string) *Tray {
	return &Tray{
		Item: Item{
			caption,
		},
		Tray: []ItemInterface{},
	}
}

func (t *Tray) Add(item ItemInterface) {
	t.Tray = append(t.Tray, item)
}

/*****************************
 * Page
 ****************************/
type PageInterface interface {
	MakeHTML() string
	Add(item ItemInterface)
	Output(filename string, page PageInterface)
}

type Page struct {
	Title   string
	Author  string
	Content []ItemInterface
}

func NewPage(title, author string) *Page {
	return &Page{
		Title:   title,
		Author:  author,
		Content: []ItemInterface{},
	}
}

func (p *Page) Add(item ItemInterface) {
	p.Content = append(p.Content, item)
}

func (p *Page) Output(filename string, page PageInterface) {
	file, err := os.Create(filename)
	if err != nil {
		log.Fatal("couldn't open file")
	}
	defer file.Close()

	byteBody := []byte(page.MakeHTML())
	_, err = file.Write(byteBody)
	if err != nil {
		log.Fatal("couldn't write body to file")
	}

	fmt.Println(filename + "を作成しました。")

}

/*****************************
 * Factory
 ****************************/
type FactoryInterface interface {
	CreateLink(caption, url string) LinkInterface
	CreateTray(caption string) TrayInterface
	CreatePage(title, author string) PageInterface
}
